### DISCUSSION TITLE
In a new processor/language definition, is it possible to change the hex sigil from `$`?

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#6992] [Q&A] [answered]
I'm experimenting with Ghidra for the old TI-99/4A, which used `>` instead of `$` as the hexadecimal prefix in both TMS-9900 assembly language and its GPL (Graphics Programming Language) assembly language, which is for a byte-code used extensively on the machine.

Is it possible to change the hexadecimal prefix string for a language?

If not, which source code file(s) in the source are responsible for rendering integers in the disassembly listing? Is it part of the Sleigh system? Or does that system just supply numbers that a higher-level component renders? I'm guessing it's in the Sleigh code since other parts of the assembly code can be rendered from string literals in the `.slaspec` files.

It's hard to search for, so perhaps somebody knows the code well enough to point in the right direction.

**Comment by GhidorahRex** [comment, marked as answer]
I assume you mean for instruction operands? We don't normally use `$` as a prefix string for hex literals. We use `0x`. Ghidra does not allow customization of this prefix (or suffix in the case of `FFFFh`).

The code responsible for the listing display is in `Scalar.java` on line 286. But I can't guarantee that changing that won't have weird side-effects, but that is where it is located.

**Reply by hippietrail** [reply]
Oops silly me. I should've double checked before writing from memory. But now I noticed that while my very experimental TI-99/4A GPL processor does show `0x` prefixes for hex, my experimental Apple Lisa loader is showing `h` suffixes for hex..

`0x` prefix:
<img width="354" alt="image" src="https://github.com/user-attachments/assets/a8ed975d-e433-4f59-a7cd-dec05bda3248">

`h` suffix: 
<img width="424" alt="image" src="https://github.com/user-attachments/assets/607ac13b-cc47-4a2e-af8d-f2aba90696f1">

Silly me again. I should've read your message more thoroughly before replying. I'll look at `Scalar.java` and thank you.

**Reply by GhidorahRex** [reply]
Data is handled differently than instruction operands (notice how the operands in the Apple Lisa are still prefixed by '0x', only the data has the suffix).

**Reply by hippietrail** [reply]
> The code responsible for the listing display is in `Scalar.java` on line 286. But I can't guarantee that changing that won't have weird side-effects, but that is where it is located.

I'm investigating what it would take to implement this. So just noting here that the two main places with code responsible for the `h` suffix are `DefaultDataType.java` and `AbstractIntegerDataType.java`. The latter is for fields of `struct`s.

**Reply by ibm5110** [reply]
> I assume you mean for instruction operands? We don't normally use `$` as a prefix string for hex literals. We use `0x`. Ghidra does not allow customization of this prefix (or suffix in the case of `FFFFh`).

Is there any progess on this subject? The prefix '0x' is so annoying in assembly languages.

**Comment by ibm5110** [comment]
I'm joining this discussion as I face the same "problem". Almost all assembly languages do _not_ use 0x (or C style) prefixes. I am writing a custom processor module and I want to set the representation of numbers, i.e. $ or X' for hexadecimal values, and other things.
I can see that in the aforementioned Scalar.java the function toString has the options of providing a prefix and a postfix (e.g. for Intel hex notation). Would it be possible to set the defaults somewhere in the GUI?

**Reply by russdill** [reply]
I would love to be able to include it in the cspec or something. Having ghidra generate assembly that existing tools can consume is really helpful for automated test.
